File: rig/da.py
"""Data-availability layer: erasure coding + availability sampling (§3.3).

The rig modelled DA as "the body travels with its tx" — withholding was caught
only because the body was present to hash. A real DA layer must let the network
be *sure a body is retrievable* without any one node downloading it, and stay
recoverable when some holders vanish or lie. This is the primitive Celestia
built and the whitepaper's §3.3 calls for.

Two pieces:

  * **Erasure coding** (Reed-Solomon over GF(256)). A delta body is split into
    k data-carrying symbols per column and expanded to n shards such that ANY k
    of the n reconstruct it. An adversary must therefore withhold more than n−k
    shards to make the body unrecoverable.
  * **Availability sampling.** The n shards are Merkle-committed; the root is the
    tx's DA pointer. A verifier requests a few random shards with Merkle proofs.
    Because unrecoverability requires >n−k shards missing, a handful of random
    samples detects a withholding attack with high probability — without ever
    fetching the whole body.

GF(256) is implemented here (≈ the AES field) so the rig stays dependency-free.
"""

import hashlib
from dataclasses import dataclass

from . import merkle

# --------------------------------------------------------------------------
# GF(256) — the AES field, generator 0x03, modulus x^8+x^4+x^3+x+1 (0x11b)
# --------------------------------------------------------------------------
_EXP = [0] * 512
_LOG = [0] * 256
_x = 1
for _i in range(255):
    _EXP[_i] = _x
    _LOG[_x] = _i
    _x ^= (_x << 1)
    if _x & 0x100:
        _x ^= 0x11b
for _i in range(255, 512):
    _EXP[_i] = _EXP[_i - 255]


def _mul(a, b):
    if a == 0 or b == 0:
        return 0
    return _EXP[_LOG[a] + _LOG[b]]


def _inv(a):
    return _EXP[255 - _LOG[a]]


def _mat_inv(m):
    """Invert a k×k GF(256) matrix by Gauss-Jordan; returns None if singular."""
    k = len(m)
    a = [row[:] + [1 if i == j else 0 for j in range(k)] for i, row in enumerate(m)]
    for col in range(k):
        piv = next((r for r in range(col, k) if a[r][col] != 0), None)
        if piv is None:
            return None
        a[col], a[piv] = a[piv], a[col]
        inv = _inv(a[col][col])
        a[col] = [_mul(v, inv) for v in a[col]]
        for r in range(k):
            if r != col and a[r][col]:
                f = a[r][col]
                a[r] = [x ^ _mul(f, y) for x, y in zip(a[r], a[col])]
    return [row[k:] for row in a]


def _vandermonde(n, k):
    """n×k Vandermonde: row i uses evaluation point (i+1); any k rows invertible."""
    return [[_EXP[(_LOG[i + 1] * j) % 255] if (i + 1) != 0 else (1 if j == 0 else 0)
             for j in range(k)] for i in range(n)]


# --------------------------------------------------------------------------
# Erasure coding
# --------------------------------------------------------------------------
def encode(body: bytes, k: int, n: int) -> list[bytes]:
    """Split `body` into k data rows and expand to n shards (any k reconstruct)."""
    assert 0 < k <= n <= 255
    pad = (-len(body)) % k
    data = body + b"\x00" * pad
    L = len(data) // k
    rows = [data[r * L:(r + 1) * L] for r in range(k)]        # k rows, L bytes each
    V = _vandermonde(n, k)
    shards = []
    for i in range(n):
        vi = V[i]
        out = bytearray(L)
        for col in range(L):
            acc = 0
            for r in range(k):
                acc ^= _mul(vi[r], rows[r][col])
            out[col] = acc
        shards.append(bytes(out))
    return shards


def reconstruct(shards: dict, k: int, orig_len: int) -> bytes:
    """Recover the body from any k shards ({index: bytes}); raises if < k."""
    if len(shards) < k:
        raise ValueError(f"need {k} shards, have {len(shards)}")
    idx = sorted(shards)[:k]
    V = _vandermonde(max(idx) + 1, k)
    sub = [V[i] for i in idx]
    inv = _mat_inv(sub)
    if inv is None:
        raise ValueError("singular shard set (should not happen for distinct rows)")
    L = len(shards[idx[0]])
    rows = []
    for r in range(k):
        row = bytearray(L)
        for col in range(L):
            acc = 0
            for c in range(k):
                acc ^= _mul(inv[r][c], shards[idx[c]][col])
            row[col] = acc
        rows.append(bytes(row))
    return b"".join(rows)[:orig_len]


# --------------------------------------------------------------------------
# DA blob: erasure-coded, Merkle-committed, samplable
# --------------------------------------------------------------------------
@dataclass
class DABlob:
    shards: list
    orig_len: int
    k: int
    n: int
    levels: list                     # merkle tree over shards

    @property
    def root(self) -> bytes:
        return self.levels[-1][0]

    def proof(self, i: int):
        return merkle.proof(self.levels, i)


def disperse(body: bytes, k: int, n: int) -> DABlob:
    shards = encode(body, k, n)
    return DABlob(shards=shards, orig_len=len(body), k=k, n=n,
                  levels=merkle.build(shards))


def verify_shard(shard: bytes, i: int, proof, root: bytes) -> bool:
    """A sampled shard is authentic iff its Merkle proof checks against root."""
    return merkle.verify(shard, i, proof, root)


def sample_available(available: dict, blob: DABlob, num_samples, rng) -> bool:
    """Availability sampling: request `num_samples` random shards; the body is
    deemed available iff every sampled shard is present AND proves against root.
    Unrecoverability needs > n−k missing, so a few samples catch withholding
    with high probability."""
    picks = rng.choice(blob.n, size=min(num_samples, blob.n), replace=False)
    for i in picks:
        i = int(i)
        if i not in available:
            return False                       # a sampled shard is missing
        if not verify_shard(available[i], i, blob.proof(i), blob.root):
            return False                       # …or forged
    return True


def detection_probability(available_count, n, k, num_samples) -> float:
    """P(random sampling hits at least one missing shard) when the body is
    unrecoverable (available_count ≤ n−k)."""
    from math import comb
    miss = n - available_count
    if miss <= 0 or num_samples > available_count:
        return 1.0
    p_all_present = comb(available_count, num_samples) / comb(n, num_samples)
    return 1.0 - p_all_present


def da_pointer(root: bytes) -> str:
    return "da://" + root.hex()[:32]


if __name__ == "__main__":
    import numpy as np
    print("=" * 70)
    print("  PALIMPSEST — data-availability layer (erasure coding + sampling)")
    print("=" * 70)
    rng = np.random.default_rng(0)
    body = rng.integers(0, 256, size=4096, dtype=np.uint8).tobytes()
    k, n = 4, 12
    blob = disperse(body, k, n)
    print(f"\ndelta body {len(body)} B  ->  {n} shards of {len(blob.shards[0])} B "
          f"(any {k} reconstruct)")
    print(f"  DA commitment (Merkle root): {da_pointer(blob.root)}")

    # reconstruct from an arbitrary k-subset
    keep = sorted(rng.choice(n, size=k, replace=False).tolist())
    got = reconstruct({i: blob.shards[i] for i in keep}, k, blob.orig_len)
    print(f"  reconstruct from shards {keep}: exact = {got == body}")

    # availability sampling on a healthy blob
    all_avail = {i: blob.shards[i] for i in range(n)}
    ok = sample_available(all_avail, blob, num_samples=4, rng=np.random.default_rng(1))
    print(f"\n  sampling a fully-available blob (4 samples): available = {ok}")

    # withholding attack: adversary serves only k-1 shards (unrecoverable — a
    # body needs any k, so fewer than k available can never be reconstructed)
    serve = k - 1
    withheld = {i: blob.shards[i] for i in range(serve)}
    caught = 0
    trials = 400
    for s in range(trials):
        if not sample_available(withheld, blob, num_samples=4,
                                rng=np.random.default_rng(1000 + s)):
            caught += 1
    p = detection_probability(serve, n, k, 4)
    print(f"  withholding attack ({serve}/{n} shards, unrecoverable):")
    print(f"    reconstruct fails: ", end="")
    try:
        reconstruct(withheld, k, blob.orig_len); print("NO (bug)")
    except ValueError:
        print("yes")
    print(f"    sampling caught it {caught}/{trials} times "
          f"(theory {p*100:.0f}% per check)")

    # a forged shard is rejected by its Merkle proof
    forged = dict(all_avail)
    forged[3] = bytes(len(blob.shards[3]))
    bad = sample_available(forged, blob, num_samples=n, rng=np.random.default_rng(7))
    print(f"\n  forged shard rejected by Merkle proof: {not bad}")
    print("=" * 70)
    ok_all = (got == body and ok and caught > trials * 0.9 and not bad)
    raise SystemExit(0 if ok_all else 1)

File: node/core/src/da.rs
//! Data-availability layer: erasure coding + availability sampling (§3.3).
//!
//! A faithful port of the Python reference (`rig/da.py` + `rig/merkle.py`) — the
//! golden vectors pin this to it byte-for-byte. A delta body is Reed-Solomon
//! coded over GF(256) into `n` shards such that ANY `k` reconstruct it, the
//! shards are Merkle-committed (the root IS the tx's DA pointer), and a verifier
//! samples a few random shards with inclusion proofs. Because unrecoverability
//! needs more than `n-k` shards missing, a handful of samples detects a
//! withholding attack with high probability — without downloading the body.

use sha2::{Digest, Sha256};
use std::collections::BTreeMap;
use std::sync::OnceLock;

// ---------------------------------------------------------------------------
// GF(256) — the AES field: generator 0x03, modulus x^8+x^4+x^3+x+1 (0x11b)
// ---------------------------------------------------------------------------

struct Gf {
    exp: [u8; 512],
    log: [u8; 256],
}

fn gf() -> &'static Gf {
    static GF: OnceLock<Gf> = OnceLock::new();
    GF.get_or_init(|| {
        let mut exp = [0u8; 512];
        let mut log = [0u8; 256];
        let mut x: u16 = 1;
        for i in 0..255 {
            exp[i] = x as u8;
            log[x as usize] = i as u8;
            x ^= x << 1;
            if x & 0x100 != 0 {
                x ^= 0x11b;
            }
        }
        for i in 255..512 {
            exp[i] = exp[i - 255];
        }
        Gf { exp, log }
    })
}

fn gf_mul(a: u8, b: u8) -> u8 {
    if a == 0 || b == 0 {
        return 0;
    }
    let g = gf();
    // log a + log b < 512, so the 512-entry exp table needs no modulo (matches
    // the Python `_EXP[_LOG[a] + _LOG[b]]`).
    g.exp[g.log[a as usize] as usize + g.log[b as usize] as usize]
}

fn gf_inv(a: u8) -> u8 {
    let g = gf();
    g.exp[255 - g.log[a as usize] as usize]
}

/// n×k Vandermonde: row i uses evaluation point (i+1); any k rows are invertible.
fn vandermonde(n: usize, k: usize) -> Vec<Vec<u8>> {
    let g = gf();
    (0..n)
        .map(|i| {
            (0..k)
                .map(|j| {
                    // V[i][j] = EXP[(LOG[i+1] * j) % 255]; i+1 >= 1 always.
                    let idx = (g.log[i + 1] as usize * j) % 255;
                    g.exp[idx]
                })
                .collect()
        })
        .collect()
}

/// Invert a k×k GF(256) matrix by Gauss-Jordan; None if singular.
fn mat_inv(m: &[Vec<u8>]) -> Option<Vec<Vec<u8>>> {
    let k = m.len();
    // augmented [m | I]
    let mut a: Vec<Vec<u8>> = m
        .iter()
        .enumerate()
        .map(|(i, row)| {
            let mut r = row.clone();
            r.extend((0..k).map(|j| if i == j { 1u8 } else { 0u8 }));
            r
        })
        .collect();
    for col in 0..k {
        let piv = (col..k).find(|&r| a[r][col] != 0)?;
        a.swap(col, piv);
        let inv = gf_inv(a[col][col]);
        for v in a[col].iter_mut() {
            *v = gf_mul(*v, inv);
        }
        for r in 0..k {
            if r != col && a[r][col] != 0 {
                let f = a[r][col];
                let pivot_row = a[col].clone();
                for (x, y) in a[r].iter_mut().zip(pivot_row.iter()) {
                    *x ^= gf_mul(f, *y);
                }
            }
        }
    }
    Some(a.into_iter().map(|row| row[k..].to_vec()).collect())
}

/// One output row = XOR_r ( coeffs[r] · byte_rows[r] ), over GF(256).
fn gf_combine(coeffs: &[u8], byte_rows: &[Vec<u8>]) -> Vec<u8> {
    let len = byte_rows[0].len();
    let mut acc = vec![0u8; len];
    for (&c, row) in coeffs.iter().zip(byte_rows) {
        if c != 0 {
            for (a, &r) in acc.iter_mut().zip(row) {
                *a ^= gf_mul(c, r);
            }
        }
    }
    acc
}

// ---------------------------------------------------------------------------
// Erasure coding
// ---------------------------------------------------------------------------

/// Split `body` into k data rows and expand to n shards (any k reconstruct).
pub fn encode(body: &[u8], k: usize, n: usize) -> Vec<Vec<u8>> {
    assert!(0 < k && k <= n && n <= 255, "require 0 < k <= n <= 255");
    let pad = (k - body.len() % k) % k;
    let mut data = body.to_vec();
    data.extend(std::iter::repeat(0u8).take(pad));
    let l = data.len() / k;
    let rows: Vec<Vec<u8>> = (0..k).map(|r| data[r * l..(r + 1) * l].to_vec()).collect();
    let v = vandermonde(n, k);
    (0..n).map(|i| gf_combine(&v[i], &rows)).collect()
}

/// Recover the body from any k shards ({index: bytes}); None if fewer than k or
/// the chosen shard rows are singular (never happens for distinct Vandermonde
/// rows).
pub fn reconstruct(shards: &BTreeMap<usize, Vec<u8>>, k: usize, orig_len: usize) -> Option<Vec<u8>> {
    if shards.len() < k {
        return None;
    }
    let idx: Vec<usize> = shards.keys().copied().take(k).collect();
    let v = vandermonde(idx.iter().max().copied().unwrap() + 1, k);
    let sub: Vec<Vec<u8>> = idx.iter().map(|&i| v[i].clone()).collect();
    let inv = mat_inv(&sub)?;
    let cols: Vec<Vec<u8>> = idx.iter().map(|i| shards[i].clone()).collect();
    let mut out = Vec::with_capacity(k * cols[0].len());
    for r in 0..k {
        out.extend(gf_combine(&inv[r], &cols));
    }
    out.truncate(orig_len);
    Some(out)
}

// ---------------------------------------------------------------------------
// Merkle commitment over shards (mirrors rig/merkle.py exactly)
// ---------------------------------------------------------------------------

fn leaf_hash(data: &[u8]) -> [u8; 32] {
    let mut m = Sha256::new();
    m.update([0u8]); // domain-separated leaf
    m.update(data);
    m.finalize().into()
}

fn node_hash(a: &[u8; 32], b: &[u8; 32]) -> [u8; 32] {
    let mut m = Sha256::new();
    m.update([1u8]); // domain-separated internal node
    m.update(a);
    m.update(b);
    m.finalize().into()
}

/// Build the tree as levels[0]=leaf hashes .. levels[last]=[root]. An odd node
/// is promoted (hashed with itself).
pub fn merkle_build(leaves: &[Vec<u8>]) -> Vec<Vec<[u8; 32]>> {
    assert!(!leaves.is_empty(), "need at least one leaf");
    let mut level: Vec<[u8; 32]> = leaves.iter().map(|x| leaf_hash(x)).collect();
    let mut levels = vec![level.clone()];
    while level.len() > 1 {
        let mut nxt = Vec::with_capacity(level.len().div_ceil(2));
        let mut i = 0;
        while i < level.len() {
            let a = level[i];
            let b = if i + 1 < level.len() { level[i + 1] } else { level[i] };
            nxt.push(node_hash(&a, &b));
            i += 2;
        }
        levels.push(nxt.clone());
        level = nxt;
    }
    levels
}

/// A proof step: sibling on the ('L'eft | 'R'ight) of the path node.
pub type ProofStep = (char, [u8; 32]);

pub fn merkle_proof(levels: &[Vec<[u8; 32]>], index: usize) -> Vec<ProofStep> {
    let mut path = Vec::new();
    let mut idx = index;
    for level in &levels[..levels.len() - 1] {
        if idx % 2 == 0 {
            let sib = if idx + 1 < level.len() { level[idx + 1] } else { level[idx] };
            path.push(('R', sib));
        } else {
            path.push(('L', level[idx - 1]));
        }
        idx /= 2;
    }
    path
}

pub fn merkle_verify(data: &[u8], path: &[ProofStep], root: &[u8; 32]) -> bool {
    let mut h = leaf_hash(data);
    for &(side, sib) in path {
        h = if side == 'L' { node_hash(&sib, &h) } else { node_hash(&h, &sib) };
    }
    &h == root
}

// ---------------------------------------------------------------------------
// DA blob + availability sampling
// ---------------------------------------------------------------------------

pub struct DaBlob {
    pub shards: Vec<Vec<u8>>,
    pub orig_len: usize,
    pub k: usize,
    pub n: usize,
    levels: Vec<Vec<[u8; 32]>>,
}

impl DaBlob {
    pub fn root(&self) -> [u8; 32] {
        self.levels[self.levels.len() - 1][0]
    }
    pub fn proof(&self, i: usize) -> Vec<ProofStep> {
        merkle_proof(&self.levels, i)
    }
}

/// Erasure-code + Merkle-commit a body into a samplable DA blob.
pub fn disperse(body: &[u8], k: usize, n: usize) -> DaBlob {
    let shards = encode(body, k, n);
    let levels = merkle_build(&shards);
    DaBlob { shards, orig_len: body.len(), k, n, levels }
}

/// The DA pointer committed on-chain: `da://` + first 32 hex chars of the root.
pub fn da_pointer(root: &[u8; 32]) -> String {
    format!("da://{}", &hex::encode(root)[..32])
}

/// Deterministic sample indices from a seed — validators must sample the SAME
/// shards to reach the same availability verdict, so this replaces the
/// reference's RNG with a seed-driven (verifiable) selection: a distinct
/// permutation prefix derived by hashing (seed, counter).
pub fn sample_indices(seed: &[u8], n: usize, num_samples: usize) -> Vec<usize> {
    let want = num_samples.min(n);
    let mut chosen: Vec<usize> = Vec::with_capacity(want);
    let mut counter: u64 = 0;
    while chosen.len() < want {
        let mut m = Sha256::new();
        m.update(seed);
        m.update(counter.to_le_bytes());
        let d: [u8; 32] = m.finalize().into();
        let pick = (u64::from_le_bytes(d[..8].try_into().unwrap()) % n as u64) as usize;
        if !chosen.contains(&pick) {
            chosen.push(pick);
        }
        counter += 1;
    }
    chosen
}

/// Availability sampling: the body is deemed available iff every sampled shard
/// is present AND proves against the blob's root. Unrecoverability needs
/// > n-k missing, so a few samples catch withholding with high probability.
pub fn sample_available(
    available: &BTreeMap<usize, Vec<u8>>,
    blob: &DaBlob,
    seed: &[u8],
    num_samples: usize,
) -> bool {
    let root = blob.root();
    for i in sample_indices(seed, blob.n, num_samples) {
        match available.get(&i) {
            None => return false, // a sampled shard is missing
            Some(shard) => {
                if !merkle_verify(shard, &blob.proof(i), &root) {
                    return false; // ...or forged
                }
            }
        }
    }
    true
}
